### PR TITLE
feat: Add Options for Slug

### LINKED ISSUE #3: Support for options (a la `slugify`)
## Environment

* `purescript` 0.15.14
* `purescript-slug` 3.0.8

## Current behavior

A `Slug` cannot contain non-alphanumeric characters nor upper-case characters, so this library will fail to parse "slugs" generated by other slug libraries (like the popular `slugify` package on npm) that _can_ contain those characters _by default_. Because of this we needed to vendor this library and modify it to fix a bug in the `halogen-realworld` example: https://github.com/thomashoneyman/purescript-halogen-realworld/issues/127.

## Expected behavior

An alternative could be to support similar configuration options that `slugify` supports: https://www.npmjs.com/package/slugify#options
```js
slugify('some string', {
  replacement: '-',  // replace spaces with replacement character, defaults to `-`
  remove: undefined, // remove characters that match regex, defaults to `undefined`
  lower: false,      // convert to lower case, defaults to `false`
  strict: false,     // strip special characters except replacement, defaults to `false`
  locale: 'vi',      // language code of the locale to use
  trim: true         // trim leading and trailing replacement chars, defaults to `true`
})
```
I haven't thought much about what this API would look like ported to PureScript. I think `locale` support might be tricky, but I suppose the others wouldn't be too difficult to implement?

Understandable if this is out of scope for this library, but if there's interest in adding this I could look into it.

### PR DESCRIPTION
**Pull request by pete-murphy**
## What does this pull request do?

Fixes #3 

This adds `Options` for configuring `generateWithOptions` and `parseWithOptions`. 

```purescript
type Options =
  { replacement :: String           -- used to replace spaces 
  , keepIf :: CodePoint -> Boolean  -- which characters are allowed in the slug
  , lowerCase :: Boolean            -- transform characters to lower-case
  , stripApostrophes :: Boolean     -- remove apostrophes (to avoid unnecessary word-breaks)
  }
```

The changes are modeled after `slugify`'s options argument:

```javascript
slugify('some string', {
  replacement: '-',  // replace spaces with replacement character, defaults to `-`
  remove: undefined, // remove characters that match regex, defaults to `undefined`
  lower: false,      // convert to lower case, defaults to `false`
  strict: false,     // strip special characters except replacement, defaults to `false`
  locale: 'vi',      // language code of the locale to use
  trim: true         // trim leading and trailing replacement chars, defaults to `true`
})
```

`replacement` appears in `Options` as well, and `lower` appears as `lowerCase`. I think the use case for  `remove` and `strict` are covered by `keepIf` (I'd welcome any suggestion for better naming). From what I can tell, `strict` was introduced to cover a common use case for `remove` which is to remove all non-alphanumeric characters: https://github.com/simov/slugify/blob/4860d7efd6734af8c716f26cd246aa5b1e503163/slugify.js#L45-L47.
In `slugify`, `remove` takes a `Regex`, but from examples I've seen it's usually just for matching on a set of characters, so the full power of `Regex` seems unnecessary. I think `CodePoint -> Boolean` works as a corresponding PureScript-y API, and also neatly takes advantage of the `HeytingAlgebra` instance for functions (so one can do `isLatin1 && isAlphaNum` etc.)

I didn't do anything related to `locale` or `trim`.

Also, the intent is that `generate` and `parse` work the same as before these changes, though they're now implemented in terms of the `*WithOptions` functions.

## Where should the reviewer start?

Start with changes to `Slug` implementations of `generate` and `parse`, along with documentation changes in there.

## Other Notes:

> Does the project documentation need to be updated? Should we introduce a new example, or update any existing examples? Does this PR require any other follow-up tasks?

Yes, and I think all relevant documentation has been updated. Note that a trade-off is the loss of guarantees: one no longer knows much about a value of type `Slug` other than that it's non-empty.

### CODE REVIEW

File: test/Main.purs [hunk 1]
```diff
@@ -76,6 +78,23 @@ main = launchAff_ $ runSpec [consoleReporter] do
...
+        f x `assertEquals` g x
+
+  describe "Parse Slugs with Options" do
+    it "Slug parses successfully on valid input" do
+      quickCheck' 500 \(SlugWithOptions { slug, options }) -> do
+        Slug.parseWithOptions options (Slug.toString slug) `assertEquals` pure slug
+
+    it "Slug fails to parse bad input" do
+      quickCheck' 500 \(BadInputWithOptions { input, options }) -> do
+        Slug.parseWithOptions options input `assertEquals` Nothing
```
Comment (pete-murphy) [inline test/Main.purs:96]: Both of these are currently failing, for example with:
```
Generate Slugs with Options
  ✗ Generated slugs are idempotent:

  (Just (Slug ⦖঍Ǐ鐵濟昒嚩씈�标ᜫ碉)) /= (Just (Slug ⦖঍Ǐ঍Ǐ鐵濟昒嚩씈�濟昒嚩঍Ǐ鐵濟昒嚩씈�标ᜫ碉))
  (Just (Slug ▄⮓饕ꭡ峀�撟᳕‬ᇟ根䵮)) /= (Just (Slug ▄⮓饕ꭡ峀�撟᳕‬ᇟ根᳕‬ᇟ饕ꭡ峀�撟᳕‬ᇟ根䵮))
  (Just (Slug ꣵ묛⍂떖犒꧑繜)) /= (Just (Slug ꣵ묛⍂떖犒꧑떖犒꧑繜))
Parse Slugs with Options
  ✗ Slug parses successfully on valid input:

  Nothing /= (Just (Slug 褬א᫓萈�ﻳ↹Ⅺ矻筷ꀊ))
  Nothing /= (Just (Slug 5啖衧㆑穦�纬勊穦�纬迏))
  Nothing /= (Just (Slug 쁦꺶衒⍄4꼼疀))
  Nothing /= (Just (Slug 섓鄷잲㦱ၩ쭀躩֤堿魌ꂞ㦱ၩ쭀躩֤堿))
  Nothing /= (Just (Slug 赚㚂䒪쏄䧀襒ढ़╦❉፣ጎ悷㼇))
  Nothing /= (Just (Slug ꨡ祡빰搅ᯝ苊ള郚怣ᕥ쨜빰搅ᯝ苊ള郚怣ᕥ릹觪))
```
I'll dig into why that's failing when I have a chance, but maybe I need to restrict `Options` to avoid edge cases here.

File: test/Main.purs [hunk 2]
```diff
@@ -76,6 +78,23 @@ main = launchAff_ $ runSpec [consoleReporter] do
...
+      quickCheck' 500 \(Options' options) x -> do
+        let generate = Slug.generateWithOptions options
+            f = generate
+            g = generate >=> generate <<< Slug.toString
+        f x `assertEquals` g x
+
+  describe "Parse Slugs with Options" do
+    it "Slug parses successfully on valid input" do
+      quickCheck' 500 \(SlugWithOptions { slug, options }) -> do
+        Slug.parseWithOptions options (Slug.toString slug) `assertEquals` pure slug
```
Comment (pete-murphy) [inline test/Main.purs:93]: Both of these are currently failing, for example with:
```
Generate Slugs with Options
  ✗ Generated slugs are idempotent:

  (Just (Slug ⦖঍Ǐ鐵濟昒嚩씈�标ᜫ碉)) /= (Just (Slug ⦖঍Ǐ঍Ǐ鐵濟昒嚩씈�濟昒嚩঍Ǐ鐵濟昒嚩씈�标ᜫ碉))
  (Just (Slug ▄⮓饕ꭡ峀�撟᳕‬ᇟ根䵮)) /= (Just (Slug ▄⮓饕ꭡ峀�撟᳕‬ᇟ根᳕‬ᇟ饕ꭡ峀�撟᳕‬ᇟ根䵮))
  (Just (Slug ꣵ묛⍂떖犒꧑繜)) /= (Just (Slug ꣵ묛⍂떖犒꧑떖犒꧑繜))
Parse Slugs with Options
  ✗ Slug parses successfully on valid input:

  Nothing /= (Just (Slug 褬א᫓萈�ﻳ↹Ⅺ矻筷ꀊ))
  Nothing /= (Just (Slug 5啖衧㆑穦�纬勊穦�纬迏))
  Nothing /= (Just (Slug 쁦꺶衒⍄4꼼疀))
  Nothing /= (Just (Slug 섓鄷잲㦱ၩ쭀躩֤堿魌ꂞ㦱ၩ쭀躩֤堿))
  Nothing /= (Just (Slug 赚㚂䒪쏄䧀襒ढ़╦❉፣ጎ悷㼇))
  Nothing /= (Just (Slug ꨡ祡빰搅ᯝ苊ള郚怣ᕥ쨜빰搅ᯝ苊ള郚怣ᕥ릹觪))
```
I'll dig into why that's failing when I have a chance, but maybe I need to restrict `Options` to avoid edge cases here.

Comment (pete-murphy) [reply test/Main.purs:93]: So I think these properties don't hold if the `replacement` string contains a space _and_ another character that passes the `keepIf` condition. Probably also if the `replacement` contains upper case characters, and `lowerCase == true`. Not sure what else I'm not thinking of. I just pushed some changes to restrict the `replacement` to dash or underscore, which passes the property tests, but I think the `Arbitrary` instance could reflect that constraint more accurately. I'll update it.

Comment (pete-murphy) [reply test/Main.purs:93]: Submitting for review as-is. I think it's reasonable to assume that if someone wants to use characters in the `replacement` string that would be filtered-out or transformed by the other settings, then some properties go out the window. Happy to update the API if we want to make `replacement` more precisely-typed, but it is convenient to have it just be `String`.

**Comment by pete-murphy** [comment]
Oop, forgot to update `README` etc, I can get to that shortly

File: src/Slug.purs [hunk 3]
```diff
@@ -48,8 +63,6 @@ instance DecodeJson Slug where
 -- | with `-` dashes, ensuring the string does not begin or end with a
 -- | dash, and ensuring there are never two dashes in a row.
 -- |
--- | Slugs are usually created for article titles and other resources
--- | which need a human-readable resource name in a URL.
```
Comment (pete-murphy) [inline src/Slug.purs:52]: I moved this to the type comment above, since I think it fits more as an explanation of "what a `Slug` is", especially now that the properties guaranteed in the documentation for `Slug` only hold for default `Options`.

**Review by thomashoneyman — APPROVED** [review]

**Comment by thomashoneyman** [comment]
Sorry for the wait, @pete-murphy, and thank you for the contribution!
